### PR TITLE
fix(clusterCache): don't miss finding live obj if obj is cluster-scoped and namespacedResources is in transition

### LINKED ISSUE #171: clusterCache.IsNamespaced() lacks locking
`c.namespacedResources` is a field that can be mutated concurrently so access must be protected with locking.

### PR DESCRIPTION
**Pull request by ncdc**
1. ~namespacedResources is accessed by multiple goroutines simultaneously
and therefore must be guarded by a mutex. Because this field is read
significantly more often than it's written, use an RWMutex.~ Removed this from this PR, as it might not be adequate in its current form to avoid deadlocks.

2. When Reconcile performs its logic to compare the desired state (target
objects) against the actual state (live objects), it looks up each live
object based on a key comprised of data from the target object: API
group, API kind, namespace, and name. While group, kind, and name will
always be accurate, there is a chance that the value for namespace is
not. If a cluster-scoped target object has a namespace (because it
incorrectly has a namespace from its source) or the namespace parameter
passed into the Reconcile method has a non-empty value (indicating a
default value to use on namespace-scoped objects that don't have it set
in the source), AND the resInfo ResourceInfoProvider has incomplete or
missing API discovery data, the call to IsNamespacedOrUnknown will
return true when the information is unknown. This leads to the key being
incorrect - it will have a value for namespace when it shouldn't. As a
result, indexing into liveObjByKey will fail. This failure results in
the reconciliation containing incorrect data: there will be a nil entry
appended to targetObjs when there shouldn't be.

Most likely addresses https://github.com/argoproj/argo-cd/issues/17440.

Most likely addresses at least 1 user who had cluster-scoped resources get wiped out in https://github.com/argoproj/argo-cd/issues/15058

### CODE REVIEW

**Comment by rumstead** [comment]
@ncdc awesome work! have you posted this on the Argo CD slack or plan to represent it at one of the community/contributor meetings?

**Comment by rumstead** [comment]
For https://github.com/argoproj/argo-cd/issues/15058, we have namespace on the Argo CD application because we deploy namespace-scoped resources as well so we definitely hit the condition.

**Comment by ncdc** [comment]
@rumstead yes, posted in https://cloud-native.slack.com/archives/C01TSERG0KZ/p1719420479620669?thread_ts=1709829854.313119&cid=C01TSERG0KZ and https://kubernetes.slack.com/archives/CBT6N1ASG/p1719421432387449 and https://kubernetes.slack.com/archives/CBT6N1ASG/p1719850472508739.

Not sure about my ability to attend a community meeting; will see.

**Review by cpanato — APPROVED** [review]
this lgtm

File: pkg/cache/cluster.go
Comment (ncdc) [reply pkg/cache/cluster.go:815]: I can adjust it to hold the look for an absolute minimum amount of time. I'll look into the cross-function deadlock possibility.

Comment (ncdc) [reply pkg/cache/cluster.go:815]: @jdroot I don't see where `c.lock` is locked in `sync()` - can you share what you're seeing?

Comment (ncdc) [reply pkg/cache/cluster.go:815]: Or if it's easier, I can remove the locking change from this PR

Comment (ncdc) [reply pkg/cache/cluster.go]: Removed the lock changes from this PR

Comment (jdroot) [reply pkg/cache/cluster.go]: Just linking to latest here, but it was the same when the change was in your PR: https://github.com/argoproj/gitops-engine/blob/master/pkg/cache/cluster.go#L846

RunAllAsync does run those in another Go Routine, but it blocks until they complete. So you would have:

- Go Routine 1 in Sync() waiting on Go routine 3 to complete while holding `namespacedResourcesLock`
- Go Routine 2 in Invalidate waiting on `namespacedResourcesLock` while holding `lock`
- Go Routine 3 waiting on `lock`

Comment (ncdc) [reply pkg/cache/cluster.go]: FYI that `lock` on L846 is not `c.lock` but a function-local mutex, if that makes a difference.

**Comment by rumstead** [comment]
Not an approver but lgtm for what it is worth

File: pkg/sync/reconcile.go
Comment (crenshaw-dev) [inline pkg/sync/reconcile.go:95]: What do you think about this rewriting?

```go
		namespaced, err := resInfo.IsNamespaced(gvk.GroupKind())
		unknownScope := err != nil

		var keysToCheck []kubeutil.ResourceKey
		// If we get an error, we don't know whether the resource is namespaced. So we need to check for both in the
		// live objects. If we don't check for both, then we risk missing the object and deleting it.
		if namespaced || unknownScope {
			keysToCheck = append(keysToCheck, kubeutil.NewResourceKey(gvk.Group, gvk.Kind, ns, obj.GetName()))
		}
		if !namespaced || unknownScope {
			keysToCheck = append(keysToCheck, kubeutil.NewResourceKey(gvk.Group, gvk.Kind, "", obj.GetName()))
		}
```

I think it's a bit easier to read.

I think it also avoids the double-check in cases where we _know_ the resource is namespaced.

This still passes the unit test.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/sync/reconcile.go
Comment (ncdc) [reply pkg/sync/reconcile.go:95]: I had considered going to the raw IsNamespaced call. I like this better - thanks for the suggestion.

Comment (ncdc) [reply pkg/sync/reconcile.go]: Updated!

**Comment by crenshaw-dev** [comment]
@ncdc before I merge, would you mind opening a PR on argo-cd pointing at the latest commit of this branch? That'll let us make sure all the tests still look good on the Argo CD side.

**Comment by ncdc** [comment]
It'll take me a little while (meetings) but I'll do it ASAP

**Comment by ncdc** [comment]
https://github.com/argoproj/argo-cd/pull/19040

**Review by crenshaw-dev — APPROVED** [review]
Thank you!

**Pull request merged**
